imap/command/search.go

```diff
@@ -0,0 +1,383 @@
+package command
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/ProtonMail/gluon/imap/parser"
+	"github.com/bradenaw/juniper/xslices"
+	"time"
+)
+
+type SearchCommand struct {
+	Charset string
+	Keys    []SearchKey
+}
+
+type SearchKey interface {
+	String() string
+	SanitizedString() string
+}
+
+func (s SearchCommand) String() string {
+	charsetStr := "NONE"
+	if len(s.Charset) != 0 {
+		charsetStr = s.Charset
+	}
+
+	return fmt.Sprintf("SEARCH CHARSET=%v %v", charsetStr, s.Keys)
+}
+
+func (s SearchCommand) SanitizedString() string {
+	charsetStr := "NONE"
+	if len(s.Charset) != 0 {
+		charsetStr = s.Charset
+	}
+
+	return fmt.Sprintf("SEARCH CHARSET=%v %v", charsetStr, xslices.Map(s.Keys, func(v SearchKey) string {
+		return v.SanitizedString()
+	}))
+}
+
+type SearchCommandParser struct {
+	charset        string
+	charsetAllowed bool
+}
+
+func (scp *SearchCommandParser) FromParser(p *parser.Parser) (Payload, error) {
+	//search          = "SEARCH" [SP "CHARSET" SP astring] 1*(SP search-key)
+	//                     ; CHARSET argument to MUST be registered with IANA
+	scp.charsetAllowed = true
+
+	var keys []SearchKey
+
+	for {
+		if !p.Check(parser.TokenTypeSP) {
+			break
+		}
+
+		key, err := scp.parseSearchKey(p)
```

Comment:
if you have parsed a charset, you should look up the decoder from IANA package and use it to decode any search terms, shouldn't you?

Reply:
I would rather deal with this while handling the search command. The important thing is that we apply the decoder before we use the search term.

Reply:
Additionally, it would be good to keep the original search parameters after parsing for debugging.